### PR TITLE
db/consistency_level: do not use result from heat weighted load balanc…

### PR DESCRIPTION
**Pull request by gleb-cloudius**
…er if it contains duplicates

Because of https://github.com/scylladb/scylladb/issues/9285 heat weighted load balancer may sometimes return same node twice. It may cause wrong data to be read or unexpected errors to be returned to a client. Since the original bug is not easy to fix and it is rare lets introduce a workaround. We will check for duplicates and will use non HWLB one if one is found.

Fixes: https://github.com/scylladb/scylladb/issues/20430

Need to be backported because may cause (transient) data corruption.

### CODE REVIEW

**Comment by gleb-cloudius** [comment]
The failure known https://github.com/scylladb/scylladb/issues/20060

**Comment by avikivity** [comment]
> db/consistency_level: do not use result from hit weighted load balanc… #20414

heat

**Comment by avikivity** [comment]
Missing Fixes tag

File: db/consistency_level.cc
Comment (avikivity) [inline db/consistency_level.cc:339]: Is it guaranteed that duplicates will be adjacent?

**Comment by gleb-cloudius** [comment]
> > db/consistency_level: do not use result from hit weighted load balanc… #20414
> 
> heat

Fixed, but you can say that it is based on cache hits, not on a heat of a cache :)

File: db/consistency_level.cc
Comment (avikivity) [reply db/consistency_level.cc:339]: Maybe we should push the workaround into miss_equalizing_combination?

Comment (gleb-cloudius) [reply db/consistency_level.cc:339]: As far as I saw where the problem is cumming from yes. @nyh?

**Comment by gleb-cloudius** [comment]
> Missing Fixes tag

It does not fix the bug, just works around it. There is an issue that it will fix, but an enterprise one: https://github.com/scylladb/scylla-enterprise/issues/3735

File: db/consistency_level.cc
Comment (gleb-cloudius) [reply db/consistency_level.cc:339]: > Maybe we should push the workaround into miss_equalizing_combination?

Why? It does not have original `live_endpoints`. It will just make it more complicated.

Comment (nyh) [reply db/consistency_level.cc:339]: @avikivity yes, the way ssample() ("Systematic sampling") works is that if an item has probability >1/K, it can be included more than once in the output list but adjacent. You can see an explanation of systematic sampling in a comment in heat_load_balance.cc.

@gleb-cloudius what miss_equalizing_combination could have done is not to revert to the original live_endpints but just to drop the extra node and come up with a new random one, just like we did with _extra. I admit it's more code.

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
> > Missing Fixes tag
> 
> It does not fix the bug, just works around it. There is an issue that it will fix, but an enterprise one: [scylladb/scylla-enterprise#3735](https://github.com/scylladb/scylla-enterprise/issues/3735)

Indeed. The bug in the low-level function remains - https://github.com/scylladb/scylladb/issues/9285. It will remain open.
But you should probably open a new open-source bug describing the user-visible **symptoms** - the ones described in that enterprise issue (namely, some failed writes because of some internal sanity-checking we do). Then you can say this patch fixes that new issue.

**Comment by gleb-cloudius** [comment]
> > > Missing Fixes tag
> > 
> > 
> > It does not fix the bug, just works around it. There is an issue that it will fix, but an enterprise one: [scylladb/scylla-enterprise#3735](https://github.com/scylladb/scylla-enterprise/issues/3735)
> 
> Indeed. The bug in the low-level function remains - #9285. It will remain open. But you should probably open a new open-source bug describing the user-visible **symptoms** - the ones described in that enterprise issue (namely, some failed writes because of some internal sanity-checking we do). Then you can say this patch fixes that new issue.

I opened https://github.com/scylladb/scylladb/issues/20430 and added it to the PR cover.

**Comment by nyh** [comment]
@gleb-cloudius do we have a test (even something you can run manually once, not a nice regression test) that encounters this bug (failed reads) and you can run it to check your fix really avoids the failed reads?
Another thing worth checking is that in the "normal" case HWLB still works as expected - after all a patch to never do HWLB would have also solved the failed reads :-) I don't know if we have any test for HWLB from that from the old days that we developed it?

**Comment by gleb-cloudius** [comment]
> @gleb-cloudius do we have a test (even something you can run manually once, not a nice regression test) that encounters this bug (failed reads) and you can run it to check your fix really avoids the failed reads?

The only thing that reproduced the problem is this: https://github.com/scylladb/scylla-enterprise/issues/3735 (an enterprise upgrade sct job). It passes with HWLB disable so I am pretty confident the workaround here will solve the problem.

> Another thing worth checking is that in the "normal" case HWLB still works as expected - after all a patch to never do HWLB would have also solved the failed reads :-) I don't know if we have any test for HWLB from that from the old days that we developed it?

We have a dtest for HWLB.

**Comment by nyh** [comment]
> The only thing that reproduced the problem is this: [scylladb/scylla-enterprise#3735](https://github.com/scylladb/scylla-enterprise/issues/3735) (an enterprise upgrade sct job). It passes with HWLB disable so I am pretty confident the workaround here will solve the problem.

I'm also "pretty confident" that your fix is correct, but it would be nice to check it in practice. I've seen more than once correct-looking patches that simply didn't fix the problem they were trying to fix. How difficult is it to ask the person who ran this test to run it again with your patch?

> > Another thing worth checking is that in the "normal" case HWLB still works as expected - after all a patch to never do HWLB would have also solved the failed reads :-) I don't know if we have any test for HWLB from that from the old days that we developed it?
> 
> We have a dtest for HWLB.

Good to know. Do you remember where it is? dtest? SCT?

**Comment by gleb-cloudius** [comment]
> > The only thing that reproduced the problem is this: [scylladb/scylla-enterprise#3735](https://github.com/scylladb/scylla-enterprise/issues/3735) (an enterprise upgrade sct job). It passes with HWLB disable so I am pretty confident the workaround here will solve the problem.
> 
> I'm also "pretty confident" that your fix is correct, but it would be nice to check it in practice. I've seen more than once correct-looking patches that simply didn't fix the problem they were trying to fix. How difficult is it to ask the person who ran this test to run it again with your patch?

Building special enterprise package is a PITA. Of course after the fix will end up in the enterprise the test will be re-ran.

> 
> > > Another thing worth checking is that in the "normal" case HWLB still works as expected - after all a patch to never do HWLB would have also solved the failed reads :-) I don't know if we have any test for HWLB from that from the old days that we developed it?
> > 
> > 
> > We have a dtest for HWLB.
> 
> Good to know. Do you remember where it is? dtest? SCT?

What do you mean? I wrote above that it is a dtest. Or do we run some dtests as part of SCT?

**Comment by nyh** [comment]
> > > We have a dtest for HWLB.
> > 
> > 
> > Good to know. Do you remember where it is? dtest? SCT?
> 
> What do you mean? I wrote above that it is a dtest. Or do we run some dtests as part of SCT?

Sorry, I missed the letter "d" in your sentence :-)

**Comment by gleb-cloudius** [comment]
So what prevents this from been merged?

**Comment by nyh** [comment]
> So what prevents this from been merged?

I was hoping somebody can test that this fix actually fixes the symptoms in that SCT job. I've seen too many times really good guesses what causes a bug, and a good-looking fix, and yet the bug wasn't fixed because this wasn't the same bug or the fix added a parameter that was accidentally set to zero (true story).

If you think it's too hard to run that SCT before merging, then:
1. It's a good pointer to @fruch (who asked @avikivity a few days ago what prevents developers from using SCT),
2. But I agree we shouldn't wait until you learn to use SCT (or SCT learns to fit your needs), and will agree to merge this patch anyway (I already reviewed and approved the code) - if you "promise" to follow up later on how that SCT job does. Just tell me that you want to do that.

**Comment by gleb-cloudius** [comment]
> 
>     2. But I agree we shouldn't wait until you learn to use SCT (or SCT learns to fit your needs), and will agree to merge this patch anyway (I already reviewed and approved the code) - if you "promise" to follow up later on how that SCT job does. Just tell me that you want to do that.

Of course. We need to close the original bug eventually (which is in enterprise and will not be closed by just committing the patch).

**Comment by fruch** [comment]
> > 
> >     2. But I agree we shouldn't wait until you learn to use SCT (or SCT learns to fit your needs), and will agree to merge this patch anyway (I already reviewed and approved the code) - if you "promise" to follow up later on how that SCT job does. Just tell me that you want to do that.
> 
> Of course. We need to close the original bug eventually (which is in enterprise and will not be closed by just committing the patch).

The issue isn't just in enterprise, it happened on enterprise when we first noticed it.

it reproduces 100% on multi-dc rolling upgrade

@vponomaryov can you help @gleb-cloudius  and show how to create a reproducer job that would also do BYO build ?

**Comment by gleb-cloudius** [comment]
> > > ```
> > > 2. But I agree we shouldn't wait until you learn to use SCT (or SCT learns to fit your needs), and will agree to merge this patch anyway (I already reviewed and approved the code) - if you "promise" to follow up later on how that SCT job does. Just tell me that you want to do that.
> > > ```
> > 
> > 
> > Of course. We need to close the original bug eventually (which is in enterprise and will not be closed by just committing the patch).
> 
> The issue isn't just in enterprise, it happened on enterprise when we first noticed it.
> 
> it reproduces 100% on multi-dc rolling upgrade
> 

Really? And where is the issue about it? Because here: https://github.com/scylladb/scylla-enterprise/issues/3735#issuecomment-1875348006 it says that the issue is not reproducible on non enterprise version. And here https://github.com/scylladb/scylla-enterprise/issues/3735#issuecomment-1871390854 it says that without upgrade it did not happen as well. Which is, knowing what the bug is, not make any sense as well. I can understand that sometime you run a test once that suppose to reproduce the bug and it fails to reproduce it, but then if you do manage to reproduce you need to update the info because leaving misleading information in the issue does not help fixing it.

**Comment by fruch** [comment]
> > > > ```
> > > > 2. But I agree we shouldn't wait until you learn to use SCT (or SCT learns to fit your needs), and will agree to merge this patch anyway (I already reviewed and approved the code) - if you "promise" to follow up later on how that SCT job does. Just tell me that you want to do that.
> > > > ```
> > > 
> > > 
> > > Of course. We need to close the original bug eventually (which is in enterprise and will not be closed by just committing the patch).
> > 
> > The issue isn't just in enterprise, it happened on enterprise when we first noticed it.
> > 
> > it reproduces 100% on multi-dc rolling upgrade
> > 
> 
> Really? And where is the issue about it? Because here: https://github.com/scylladb/scylla-enterprise/issues/3735#issuecomment-1875348006 it says that the issue is not reproducible on non enterprise version. And here https://github.com/scylladb/scylla-enterprise/issues/3735#issuecomment-1871390854 it says that without upgrade it did not happen as well. Which is, knowing what the bug is, not make any sense as well. I can understand that sometime you run a test once that suppose to reproduce the bug and it fails to reproduce it, but then if you do manage to reproduce you need to update the info because leaving misleading information in the issue does not help fixing it.
> 

You are arguing about information from 8 months ago.

The full case with upgrade is getting reproduced 100% for the last 8 months.

Regardless of the trials people were doing to help narrow it down

Good luck with pushing fixes, and skipping testing. Yet again.

**Comment by gleb-cloudius** [comment]
> > > > > ```
> > > > > 2. But I agree we shouldn't wait until you learn to use SCT (or SCT learns to fit your needs), and will agree to merge this patch anyway (I already reviewed and approved the code) - if you "promise" to follow up later on how that SCT job does. Just tell me that you want to do that.
> > > > > ```
> > > > 
> > > > 
> > > > Of course. We need to close the original bug eventually (which is in enterprise and will not be closed by just committing the patch).
> > > 
> > > 
> > > The issue isn't just in enterprise, it happened on enterprise when we first noticed it.
> > > it reproduces 100% on multi-dc rolling upgrade
> > 
> > 
> > Really? And where is the issue about it? Because here: [scylladb/scylla-enterprise#3735 (comment)](https://github.com/scylladb/scylla-enterprise/issues/3735#issuecomment-1875348006) it says that the issue is not reproducible on non enterprise version. And here [scylladb/scylla-enterprise#3735 (comment)](https://github.com/scylladb/scylla-enterprise/issues/3735#issuecomment-1871390854) it says that without upgrade it did not happen as well. Which is, knowing what the bug is, not make any sense as well. I can understand that sometime you run a test once that suppose to reproduce the bug and it fails to reproduce it, but then if you do manage to reproduce you need to update the info because leaving misleading information in the issue does not help fixing it.
> 
> You are arguing about information from 8 months ago.

That's the modst up to date information I ever got. 

> 
> The full case with upgrade is getting reproduced 100% for the last 8 months.

Where is the issue. Was I pinged on it?

> 
> Regardless of the trials people were doing to help narrow it down
> 
> Good luck with pushing fixes, and skipping testing. Yet again.

Got luck with testing  and not sharing a result.

**Comment by vponomaryov** [comment]
> > > ```
> > > 2. But I agree we shouldn't wait until you learn to use SCT (or SCT learns to fit your needs), and will agree to merge this patch anyway (I already reviewed and approved the code) - if you "promise" to follow up later on how that SCT job does. Just tell me that you want to do that.
> > > ```
> > 
> > 
> > Of course. We need to close the original bug eventually (which is in enterprise and will not be closed by just committing the patch).
> 
> The issue isn't just in enterprise, it happened on enterprise when we first noticed it.
> 
> it reproduces 100% on multi-dc rolling upgrade
> 
> @vponomaryov can you help @gleb-cloudius and show how to create a reproducer job that would also do BYO build ?

Running an SCT test with custom scylla from any GH dev branch is only about setting jenkins pipeline parameters.
There is presentation about it: https://docs.google.com/presentation/d/1P5xofncoTkUI-uQ5ilG9eRhRoEfYVdJRhdps4qQ6Mx0/edit#slide=id.g2e83887fc9a_0_0

To run an SCT rolling upgrade CI job with BYO need to do following with pipeline parameters:
- remove any value from the `new_scylla_repo` pipeline param. It is what will be populated by the SCT if BYO gets used.
- define explicitly the `base_versions` pipeline parameter to some 1 version. It is a must.
- specify `byo_scylla_repo` value as your own GH repo git path, example -> `git@github.com:gleb-cloudius/scylla-enterprise.git`
- specify `byo_scylla_branch` value as `9285-workaround`.
- specify `byo_job_path` - proper path to a CI job. Example for enterprise-master - `/scylla-enterprise/byo/byo_enterprise_build_tests_dtest/`. Note that each scylla branch has it's dedicated BYO CI job.
- specify `byo_default_product` -> `scylla` or `scylla-enterprise`. Depends on the base version of your dev branch.
- specify `byo_default_branch` -> branch to be used by rel-eng projects like `scylla-pkg` and other. For scylla-enterprise it is `enterprise` or `next` for OSS-master.

Summary: any `longevity`, `rolling-upgrade` and `performance` SCT jobs can be run with scylla from any (unmerged PR)/(dev branch).

**Comment by mykaul** [comment]
> To run an SCT rolling upgrade CI job with BYO need to do following with pipeline parameters

The various fields descriptions in this job could use some more work and examples. This would make it easier and less error prone.

**Comment by gleb-cloudius** [comment]
> > > > ```
> > > > 2. But I agree we shouldn't wait until you learn to use SCT (or SCT learns to fit your needs), and will agree to merge this patch anyway (I already reviewed and approved the code) - if you "promise" to follow up later on how that SCT job does. Just tell me that you want to do that.
> > > > ```
> > > 
> > > 
> > > Of course. We need to close the original bug eventually (which is in enterprise and will not be closed by just committing the patch).
> > 
> > 
> > The issue isn't just in enterprise, it happened on enterprise when we first noticed it.
> > it reproduces 100% on multi-dc rolling upgrade
> > @vponomaryov can you help @gleb-cloudius and show how to create a reproducer job that would also do BYO build ?
> 
> Running an SCT test with custom scylla from any GH dev branch is only about setting jenkins pipeline parameters. There is presentation about it: https://docs.google.com/presentation/d/1P5xofncoTkUI-uQ5ilG9eRhRoEfYVdJRhdps4qQ6Mx0/edit#slide=id.g2e83887fc9a_0_0
> 
> To run an SCT rolling upgrade CI job with BYO need to do following with pipeline parameters:
> 
>     * remove any value from the `new_scylla_repo` pipeline param. It is what will be populated by the SCT if BYO gets used.
> 
>     * define explicitly the `base_versions` pipeline parameter to some 1 version. It is a must.
> 
>     * specify `byo_scylla_repo` value as your own GH repo git path, example -> `git@github.com:gleb-cloudius/scylla-enterprise.git`
> 
>     * specify `byo_scylla_branch` value as `9285-workaround`.
> 
>     * specify `byo_job_path` - proper path to a CI job. Example for enterprise-master - `/scylla-enterprise/byo/byo_enterprise_build_tests_dtest/`. Note that each scylla branch has it's dedicated BYO CI job.
> 
>     * specify `byo_default_product` -> `scylla` or `scylla-enterprise`. Depends on the base version of your dev branch.
> 
>     * specify `byo_default_branch` -> branch to be used by rel-eng projects like `scylla-pkg` and other. For scylla-enterprise it is `enterprise` or `next` for OSS-master.
> 
> 
> Summary: any `longevity`, `rolling-upgrade` and `performance` SCT jobs can be run with scylla from any (unmerged PR)/(dev branch).

I am sure everything above makes perfect sense to you since you do it every single day. For me nothing but setting my repo and branch makes much sense and ideally this is the only things I would need to set and choose an SCT job I want to run with it from a drop down menu. If this is an upgrade job after choosing relevant SCT it will ask me a branch to upgrade from (and BTW the bug should be reproducible with rolling restart only. If it does not we lack something in testing).

**Comment by vponomaryov** [comment]
> I am sure everything above makes perfect sense to you since you do it every single day. For me nothing but setting my repo and branch makes much sense and ideally this is the only things I would need to set and choose an SCT job I want to run with it from a drop down menu. If this is an upgrade job after choosing relevant SCT it will ask me a branch to upgrade from (and BTW the bug should be reproducible with rolling restart only. If it does not we lack something in testing).

SCT utilizes the existing BYO Scylladb CI jobs. Those jobs are scylla branch-specific.
Neither dev branch name nor dev repo say which CI job should be picked up and what is the base branch.

So, here we just utilize existing BYO ScyllaDB interface, no more, no less.

Also, specific SCT branches like `2024.2` and other have proper defaults for those additional branches.
So, in majority of the cases you will need to specify exactly 2 options which are a must for sure - `dev repo` and `dev branch`.
The only exception is the `master` branch, which is suitable for both ScyllaDB dev versions - `OSS` and `enterprise`.

**Comment by dimakr** [comment]
The fix is said to address the https://github.com/scylladb/scylla-enterprise/issues/3735 enterprise issue.
But we don't have a backport to 2024.1 and the issue does reproduce on 2024.1.10, in the latest run of [rolling-upgrade-multidc-test](https://argus.scylladb.com/test/1d5997c3-2ea5-47ca-bba5-c4edef06979a/runs?additionalRuns[]=ae79c9d5-995c-49b4-bcd5-1cd839c542f1) on 2024-09-18.
@mykaul We probably should trigger backporting the fix to enterprise 2024.1?

**Comment by mykaul** [comment]
> The fix is said to address the [scylladb/scylla-enterprise#3735](https://github.com/scylladb/scylla-enterprise/issues/3735) enterprise issue. But we don't have a backport to 2024.1 and the issue does reproduce on 2024.1.10, in the latest run of [rolling-upgrade-multidc-test](https://argus.scylladb.com/test/1d5997c3-2ea5-47ca-bba5-c4edef06979a/runs?additionalRuns%5B%5D=ae79c9d5-995c-49b4-bcd5-1cd839c542f1) on 2024-09-18. @mykaul We probably should trigger backporting the fix to enterprise 2024.1?

The fix will arrive as a result of the usual OSS -> Enterprise merge. Specifically, the 6.0 backport (https://github.com/scylladb/scylladb/commit/6bd8c9fae5551aa498553a25eeab5b145b52f558 ) will get to Enterprise.

**Comment by gleb-cloudius** [comment]
The path should apply as is. Can mergify be asked to create it. I remember there is a magic bot that gets requests.

**Comment by mykaul** [comment]
> The path should apply as is. Can mergify be asked to create it. I remember there is a magic bot that gets requests.

I'm not sure. Not yet, at least. @gleb-cloudius - can you please create a patch for 2024.1.x?

**Comment by gleb-cloudius** [comment]
> > The path should apply as is. Can mergify be asked to create it. I remember there is a magic bot that gets requests.
> 
> I'm not sure. Not yet, at least. @gleb-cloudius - can you please create a patch for 2024.1.x?

https://github.com/scylladb/scylla-enterprise/pull/4800

**Pull request closed without merging**
